### PR TITLE
reduce build image layers

### PR DESCRIPTION
**Pull request by arroyc**
This is an initial experiment on reducing layers of build images ...

### CODE REVIEW

File: images/build/Dockerfiles/Dockerfile [hunk 1]
```diff
@@ -19,16 +19,16 @@ RUN apt-get update \
     && apt-get install -y --no-install-recommends \
         # Required for .NET Core 1.1
         libunwind8 \
-    && rm -rf /var/lib/apt/lists/*
-
+    && rm -rf /var/lib/apt/lists/* \
```
Comment (kichalla) [inline images/build/Dockerfiles/Dockerfile:22]: How many layers does the full build image have after this change?

File: images/build/Dockerfiles/Dockerfile [hunk 2]
```diff
@@ -219,47 +228,41 @@ ENV GIT_COMMIT=${GIT_COMMIT}
 ENV BUILD_NUMBER=${BUILD_NUMBER}
 ENV RELEASE_TAG_NAME=${RELEASE_TAG_NAME}
 
-RUN ./build.sh dotnetcore /opt/startupcmdgen/dotnet
-RUN ./build.sh node       /opt/startupcmdgen/nodejs
-RUN ./build.sh php        /opt/startupcmdgen/php
-RUN ./build.sh python     /opt/startupcmdgen/python
+RUN ./build.sh dotnetcore /opt/startupcmdgen/dotnet \
+ && ./build.sh node       /opt/startupcmdgen/nodejs \
```
Comment (kichalla) [inline images/build/Dockerfiles/Dockerfile:232]: Did you try the idea of having a new final stage into which you copy the stuff from the current final one? I was hoping that to have much less changes compared to this one. Did you come across any issues? one issue which I can imagine is installation of system packages and pre-reqs...the issue is their installations would install in system directories which we do not know about...so I was thinking that those kind of things can be installed directly under the final-final stage and rest of the stuff could be copied from current final to the final-final stage

Comment (arroyc) [reply images/build/Dockerfiles/Dockerfile:232]: @kichalla ... if you follow from line 239, I copied stuff from node-install, python-install etc. Current build image (in master) has 96 layers and this build image has 65 layers .. I think some more layers can be optimized, but i dont think it's possible to reduce under 50.

Comment (arroyc) [reply images/build/Dockerfiles/Dockerfile:232]: > Following  are the layers ... bold lines are the copy statements from line 239 in dockerfile

d691c57cc324        6 minutes ago       /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
--
dd0edcd0f78a        6 minutes ago       /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
4b5508063014        6 minutes ago       /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
89d3978b987b        6 minutes ago       /bin/sh -c #(nop)  ARG RELEASE_TAG_NAME=unsp…   0B
79df11d49a88        6 minutes ago       /bin/sh -c #(nop)  ARG BUILD_NUMBER=unspecif…   0B
a9228c2818d7        6 minutes ago       /bin/sh -c #(nop)  ARG GIT_COMMIT=unspecified   0B
39ea8e8bd550        6 minutes ago       /bin/sh -c #(nop) COPY   dir:a103e89e1b0e2a520…   343MB
dc2da13a07f4        8 minutes ago       /bin/sh -c #(nop) COPY   dir:012a80420511d6900…   48.3MB
1258939f5071        8 minutes ago       \|2 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   107kB
**0781ad1e1318        8 minutes ago       /bin/sh -c #(nop) COPY   dir:75db6c805aaeaab25…   235MB
0e93759aadee        8 minutes ago       /bin/sh -c #(nop) COPY   dir:6a8fb16fa06067375…   623MB
856a6124d6cf        10 minutes ago      /bin/sh -c #(nop) COPY   dir:da33178b301e160f7…   368MB
38875b260abe        11 minutes ago      /bin/sh -c #(nop) COPY   dir:4c7e73895055940a9…   1.63GB
ae53e48c146d        13 minutes ago      /bin/sh -c #(nop) COPY   dir:e69c72f5ac66d9d56…   2.47GB**
c2548e9e8777        20 minutes ago      /bin/sh -c #(nop) WORKDIR /                     0B
b4f70042b0c6        26 minutes ago      /bin/sh -c #(nop)  ENV PYTHONIOENCODING=UTF-8   0B
15f3c6b4f1b6        34 minutes ago      /bin/sh -c #(nop)  ARG IMAGES_DIR               0B
8c3337b95c92        34 minutes ago      /bin/sh -c #(nop)  ARG BUILD_DIR                0B
4fc55e46ab27        34 minutes ago      \|2 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   112kB
e8a4fe9496cf        34 minutes ago      /bin/sh -c #(nop) ADD   dir:2c936622cea913b603…   406kB
ec412461fe16        34 minutes ago      /bin/sh -c #(nop) ADD   dir:95140f22d40018738c…   89.8kB
2dbb8da4bcc7        34 minutes ago      \|2 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   2.28MB
f34b1d7058f8        34 minutes ago      /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B
601f1fc7f286        34 minutes ago      /bin/sh -c #(nop)  ARG BUILD_DIR                0B
026c471b9630        34 minutes ago      /bin/sh -c #(nop)  ARG IMAGES_DIR               0B
ea909c73dcd5        35 minutes ago      /bin/sh -c #(nop)  ENTRYPOINT ["benv"]          0B
c838bd936a23        35 minutes ago      /bin/sh -c #(nop)  ENV ORYX_SDK_STORAGE_BASE…   0B
f59f3fb574b0        35 minutes ago      /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
085642a8a685        35 minutes ago      /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
e5aac720511b        35 minutes ago      /bin/sh -c #(nop)  LABEL com.microsoft.oryx.…   0B
de61993106fe        35 minutes ago      /bin/sh -c #(nop)  ARG RELEASE_TAG_NAME=unsp…   0B
ee5d573371d6        35 minutes ago      /bin/sh -c #(nop)  ARG BUILD_NUMBER=unspecif…   0B
21cb8f739335        35 minutes ago      /bin/sh -c #(nop)  ARG GIT_COMMIT=unspecified   0B
060a9c140984        35 minutes ago      /bin/sh -c #(nop)  ENV ORYX_AI_INSTRUMENTATI…   0B
2bf2921e975d        35 minutes ago      /bin/sh -c #(nop)  ARG AI_KEY                   0B
e0c3703d828d        35 minutes ago      \|4 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   10.5kB
94f15bc1ff26        35 minutes ago      /bin/sh -c #(nop) COPY   dir:ffdf7f3ff39f0c9af…   82.8MB
7d7a5b54bdff        35 minutes ago      /bin/sh -c #(nop) COPY   dir:2f6dcf44de9089a29…   255MB
3d65759db1a5        36 minutes ago      /bin/sh -c #(nop) COPY   dir:7d31f1e79c834cdc1…   54.2MB
13b537685739        36 minutes ago      /bin/sh -c #(nop) COPY   dir:5b32288f6d036fdcc…   347MB
5daadbd95d53        37 minutes ago      /bin/sh -c #(nop) COPY   dir:71673d98c0551e29a…   284MB
a6421a70bffa        37 minutes ago      /bin/sh -c #(nop) COPY   dir:51339bf2f083fee9b…   690MB
6a19dd9a582f        38 minutes ago      /bin/sh -c #(nop)  ENV NUGET_XMLDOC_MODE=ski…   0B
06d1e92c6dce        38 minutes ago      /bin/sh -c #(nop) COPY   file:bfab44bdd84a69e0…   10.4kB
c2e0b9d14b74        38 minutes ago      /bin/sh -c #(nop)  ENV PATH=/opt/oryx:/opt/n…   0B
360696e40824        38 minutes ago      /bin/sh -c #(nop)  ENV ORYX_PATHS=/opt/oryx:…   0B
6f1323f5455f        38 minutes ago      /bin/sh -c #(nop)  ENV ORIGINAL_PATH=/usr/lo…   0B
c1a154fdaedb        38 minutes ago      /bin/sh -c #(nop) WORKDIR /                     0B
ec1ed75bcee1        38 minutes ago      /bin/sh -c #(nop)  ARG SDK_STORAGE_BASE_URL_…   0B
e606db76c6f6        38 minutes ago      /bin/sh -c #(nop)  ARG SDK_STORAGE_ENV_NAME     0B
373ecfbd82d1        45 minutes ago      /bin/sh -c #(nop)  ARG IMAGES_DIR               0B
c091ef5a3310        45 minutes ago      /bin/sh -c #(nop)  ARG BUILD_DIR                0B
bc7237861823        45 minutes ago      \|2 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   112kB
1f87f3f1eb5e        45 minutes ago      /bin/sh -c #(nop) ADD   dir:2c936622cea913b603…   406kB
e6a50c240179        45 minutes ago      /bin/sh -c #(nop) ADD   dir:95140f22d40018738c…   89.8kB
676eabc8c6c5        45 minutes ago      \|2 BUILD_DIR=/tmp/oryx/build   IMAGES_DIR=/tmp…   148MB
212cd30a4db3        51 minutes ago      /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B
0349fb06fc80        51 minutes ago      /bin/sh -c #(nop)  ARG IMAGES_DIR               0B
6e869e05869c        51 minutes ago      /bin/sh -c #(nop)  ARG BUILD_DIR                0B
1f755ddd86e2        6 weeks ago         /bin/sh -c set -ex;  apt-get update;  DEBIAN…     562MB
<missing>           6 weeks ago         /bin/sh -c apt-get update &&   apt-get install…   142MB
<missing>           6 weeks ago         /bin/sh -c set -ex;  if ! command -v gpg > /…   7.81MB
<missing>           6 weeks ago         /bin/sh -c apt-get update &&   apt-get install…   23.2MB
<missing>           6 weeks ago         /bin/sh -c #(nop)  CMD ["bash"]                 0B
<missing>           6 weeks ago         /bin/sh -c #(nop) ADD   file:6e16bc9cf28c19a1f…   101MB

**Comment by kichalla** [comment]
Could you run a Nightly build too?

File: images/build/Dockerfiles/Dockerfile (hunk 1 above)
Comment (arroyc) [reply images/build/Dockerfiles/Dockerfile:22]: 60

**Comment by arroyc** [comment]
closed on behalf of #708

**Pull request closed without merging**